agent/verify_job.go

```diff
@@ -0,0 +1,82 @@
+package agent
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/buildkite/agent/v3/internal/pipeline"
+)
+
+var ErrNoSignature = errors.New("job had no signature to verify")
+
+type invalidSignatureError struct {
+	underlying error
+}
+
+func newInvalidSignatureError(err error) *invalidSignatureError {
+	return &invalidSignatureError{underlying: err}
+}
+
+func (e *invalidSignatureError) Error() string {
+	return fmt.Sprintf("invalid signature: %v", e.underlying)
+}
+
+func (e *invalidSignatureError) Unwrap() error {
+	return e.underlying
+}
+
+func (r *JobRunner) verificationFailureLogs(err error, behavior string) {
+	label := "WARNING"
+	if behavior == VerificationBehaviourBlock {
+		label = "ERROR"
+	}
+
+	r.logger.Warn("Job verification failed: %s", err.Error())
+	r.logStreamer.Process([]byte(fmt.Sprintf("⚠️ %s: Job verification failed: %s\n", label, err.Error())))
+
+	if behavior == VerificationBehaviourWarn {
+		r.logger.Warn("Job will be run without verification - this is not recommended. You can change this behavior with the `job-verification-no-signature-behavior` agent configuration option.")
+		r.logStreamer.Process([]byte(fmt.Sprintf("⚠️ %s: Job will be run without verification\n", label)))
+	}
+}
+
+func (r *JobRunner) verifyJob(verifier pipeline.Verifier) error {
+	step := r.conf.Job.Step
+
+	if step.Matrix != nil {
+		r.logger.Warn("Signing/Verification of matrix jobs is not currently supported")
+		r.logger.Warn("Watch this space 👀")
+
+		return nil
+	}
+
+	if step.Signature == nil {
+		return ErrNoSignature
+	}
+
+	// Verify the signature
+	if err := step.Signature.Verify(&step, verifier); err != nil {
+		return newInvalidSignatureError(err)
+	}
+
+	// Now that the signature of the job's step is verified, we need to check if the fields on the job match those on the
+	// step. If they don't, we need to fail the job
```

Comment:
I was going to suggest it would be simpler to override `BUILDKITE_COMMAND` with `step.Command`, than to carefully compare each field.

But thinking about it, they would only differ if the backend (ours or the hypothetical attacker's) is doing crimes, in which case we probably shouldn't run anything.

Might be worth documenting that?

Reply:
just to double check, are saying that it might be good to document that the only reason that step and job fields should differ is if an attacker is doing crimes on the backend?

Reply:
It might not be the only reason, but it definitely seems like a reason